dask/_task_spec.py

```diff
@@ -0,0 +1,941 @@
+):
+    if not all_keys:
+        all_keys = set(dsk)
+    new_dsk = {}
+    if cache is None:
+        cache = {}
+    for k, arg in dsk.items():
+        t = convert_legacy_task(k, arg, all_keys, only_refs, cache=cache)
+        if not only_refs and isinstance(t, Alias) and t.key == k:
+            continue
+        new_dsk[k] = t
+    new_dsk2 = {
+        k: v if isinstance(v, BaseTask) else LiteralTask(k, v)
+        for k, v in new_dsk.items()
+    }
+    return new_dsk2
+
+
+def resolve_aliases(dsk: dict, keys: set, dependents: dict) -> dict:
+    """Remove trivial sequential alias chains
+
+    Example:
+
+        dsk = {'x': 1, 'y': Alias('x'), 'z': Alias('y')}
+
+        resolve_aliases(dsk, {'z'}, {'x': {'y'}, 'y': {'z'}}) == {'z': 1}
+
+    """
+    if not keys:
+        raise ValueError("No keys provided")
+    dsk = dict(dsk)
+    work = list(keys)
+    seen = set()
+    while work:
+        k = work.pop()
+        if k in seen or k not in dsk:
+            continue
+        seen.add(k)
+        t = dsk[k]
+        if isinstance(t, Alias):
+            target_key = t.target.key
+            # Rules for when we allow to collapse an alias
+            # 1. The target key is not in the keys set. The keys set is what the
+            #    user is requesting and by collapsing we'd no longer be able to
+            #    return that result.
+            # 2. The target key has only one dependent which is the key we're
+            #    currently looking at. This means that there is a one to one
+            #    relation between this and the target key in which case we can
+            #    collapse them.
+            #    Note: If target was an alias as well, we could continue with
+            #    more advanced optimizations but this isn't implemented, yet
+            if (
+                target_key not in keys
+                # Note: whenever we're performing a collapse, we're not updating
+                # the dependents. The length == 1 should still be sufficient for
+                # chains of these aliases
+                and len(dependents[target_key]) == 1
+            ):
+                tnew = dsk.pop(target_key).copy()
+
+                dsk[k] = tnew
+                tnew.key = k
+                if isinstance(tnew, Alias):
+                    work.append(k)
+                    seen.discard(k)
+
+        work.extend(t.dependencies)
+    return dsk
+
+
+class KeyRef:
+    val: KeyType
+    __slots__ = ("key",)
+
+    def __init__(self, key: KeyType):
+        self.key = key
+
+    def __str__(self):
+        return str(self.key)
+
+    def __repr__(self):
+        return f"Key({repr(self.key)})"
+
+    def __eq__(self, value: object) -> bool:
+        if isinstance(value, KeyRef):
+            return self.key == value.key
+        return self.key == value
+
+    def __hash__(self) -> int:
+        return hash(self.key)
+
+
+_func_cache: MutableMapping = {}
+_func_cache_reverse: MutableMapping = {}
```

Comment:
Note this cache is unbounded. I didn't want to use an LRU (also because the acutally usable class for this is in distributed and the functools cache, etc. is not good). Also it's unclear how large the LRU should be.

Ideally, this was a cache that would work on bytes size so we could use zict, etc. for this but I didn't want to deal with this complexity just yet.